AppKit/CPWindow/CPWindow.j

```diff
@@ -3844,6 +3819,393 @@ var interpolate = function(fromValue, toValue, progress)
 
 @end
 
+@implementation CPWindow (TrackingAreaAdditions)
+
+- (void)_addTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _addTrackingArea:trackingAreas[i]];
+}
+
+- (void)_removeTrackingAreaView:(CPView)aView
+{
+    var trackingAreas = [aView trackingAreas];
+    
+    for (var i = 0; i < trackingAreas.length; i++)
+        [self _removeTrackingArea:trackingAreas[i]];
+}
+
+- (void)_addTrackingArea:(CPTrackingArea)trackingArea
+{
+    var trackingAreaView = [trackingArea view];
+    
+    if (![_trackingAreaViews containsObject:trackingAreaView])
+        [_trackingAreaViews addObject:trackingAreaView];
+    
+    // If CPTrackingAssumeInside option is set, put the tracking area in the _mouseEnteredStack
+    
+    if ([trackingArea options] & CPTrackingAssumeInside)
+        [_mouseEnteredStack addObject:trackingArea];
+}
+
+- (void)_removeTrackingArea:(CPTrackingArea)trackingArea
+{
+    // If mouse is in the tracking area, we remove it from the stack to avoid to fire a future mouseExited event
+    
+    [_mouseEnteredStack removeObjectIdenticalTo:trackingArea];
+    
+    var trackingAreaView = [trackingArea view];
+    
+    [_trackingAreaViews removeObjectIdenticalTo:trackingAreaView];
+}
+
+- (void)_handleTrackingAreaEvent:(CPEvent)anEvent
```

Comment:
Actually, for optimizations it would better to save the length of the array before entering the loop ! Otherwise each time you loop you will ask the length of the array.

`var length = array.length;
for (var i = 0; i < length; i++)`